### DISCUSSION TITLE
Running background tasks after a set amount of time.

### DISCUSSION THREAD

**Discussion by vishal-ramesh2** [fastapi/fastapi#9335] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
#
```

### Description

I have a use case where when I get a request and certain conditions are met, I need to wait T seconds and then make an API call. I don't want this to block other requests in the application.

Some of the options I have considered are, defining a function that sleeps for T seconds and then makes the API call. This function is then called using the FastAPI's `BackgroundTasks` or `asyncio.run_in_executor`.

I want to know if there is a better approach to doing this.

I do not want to use a separate application like Celery to do this because realistically, this will happen only once a day.

### Operating System

Linux

### FastAPI Version

0.92.0

### Python Version

3.10.9

**Comment by y-mathur-i** [comment]
Maybe take a look at:- https://fastapi.tiangolo.com/tutorial/background-tasks/
If you want to fire an forget the task. .i.e async execution works for you.

**Comment by iudeen** [comment]
May not be the cleanest approach, you can try something like this:

```python
import asyncio

from fastapi import FastAPI, BackgroundTasks

app = FastAPI()


# coroutine that will start another coroutine after a delay in seconds
async def delay(coro, seconds):
    # suspend for a time limit in seconds
    await asyncio.sleep(seconds)
    # execute the other coroutine
    await coro


# task coroutine to do something
async def task_coro():
    # report a message
    print('Coroutine is running')
    # simulate some long running task
    await asyncio.sleep(3)
    # report another message
    print('Coroutine is done')


@app.get("/delay/{sec}")
async def create(sec: int, bg: BackgroundTasks):
    print("received")
    bg.add_task(delay, task_coro(), sec)
```

**Comment by tahayk** [comment]
@vishal-ramesh2 
probably you can do something like this using Background tasks and [APScheduler](https://apscheduler.readthedocs.io/en/3.x/index.html)

```python
from contextlib import asynccontextmanager
from datetime import datetime, timedelta
from apscheduler.schedulers.asyncio import AsyncIOScheduler
from apscheduler.triggers.date import DateTrigger
from fastapi import FastAPI, BackgroundTasks


@asynccontextmanager
async def lifespan(app: FastAPI):
    # Startup
    print(">>>>> starting up <<<<<")

    app.schedule = AsyncIOScheduler()
    app.schedule.start()

    # App listening
    yield

    # Shutdown
    print(">>>>> shutting down <<<<<")
    app.schedule.shutdown(wait=True)


app = FastAPI(lifespan=lifespan)


def call_api(a: str):
    pass


async def delay_call(a: str):
    app.schedule.add_job(func=call_api, args=[a], trigger=DateTrigger(run_date=datetime.now() + timedelta(seconds=200)),
                         misfire_grace_time=5 * 60)

    print(">Scheduled jobs:")
    for job in app.schedule.get_jobs():
        print({"Id": str(job.id), "Run Frequency": str(job.trigger), "Next Run": str(job.next_run_time)})


@app.get("/hello")
async def get_hello(background_tasks: BackgroundTasks):
    background_tasks.add_task(delay_call, a="example arg")
    return {"data": True}
```

**Comment by vishal-ramesh2** [comment]
Thanks for the suggestions @y-mathur-i,  @iudeen, and @tahayk.

This is the solution I went with
```python
def wait_and_stop_job():
    time.sleep(T)
    ...



loop = asyncio.get_running_loop()
loop.run_in_executor(None, wait_and_stop_job)
```
